From 207604c0b5a5f22562285b889f8687a6bc9a272b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ond=C5=99ej=20Kuzn=C3=ADk?= <ondra@mistotebe.net>
Date: Thu, 7 Jul 2022 21:31:03 +0100
Subject: [PATCH] ITS#9823 Only request minCSN if accesslog is around

---
 servers/slapd/overlays/syncprov.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/servers/slapd/overlays/syncprov.c b/servers/slapd/overlays/syncprov.c
index f6ab09949..5fc39bf17 100644
--- a/servers/slapd/overlays/syncprov.c
+++ b/servers/slapd/overlays/syncprov.c
@@ -3336,7 +3336,7 @@ no_change:	if ( !(op->o_sync_mode & SLAP_SYNC_PERSIST) ) {
 					numcsns, sids, &mincsn, minsid ) ) {
 				do_present = SS_PRESENT;
 			}
-		} else if ( si->si_nopres && si->si_usehint ) {
+		} else if ( ad_minCSN != NULL && si->si_nopres && si->si_usehint ) {
 			/* We are instructed to trust minCSN if it exists. */
 			Entry *e;
 			Attribute *a = NULL;
@@ -3829,7 +3829,10 @@ sp_cf_gen(ConfigArgs *c)
 		break;
 	case SP_USEHINT:
 		si->si_usehint = c->value_int;
-		rc = syncprov_setup_accesslog();
+		if ( si->si_usehint ) {
+			/* Consider we might be a delta provider, but it's ok if not */
+			(void)syncprov_setup_accesslog();
+		}
 		break;
 	case SP_LOGDB:
 		if ( si->si_logs ) {
-- 
2.33.0

