From 0806f69c8c6ac6946c88356f65d9120bf43bdfa4 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ond=C5=99ej=20Kuzn=C3=ADk?= <okuznik@symas.com>
Date: Thu, 10 Feb 2022 17:30:17 +0000
Subject: [PATCH] ITS#9799 Clear c_n_ops_pending after we've flushed
 c_pending_ops

---
 servers/slapd/connection.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/servers/slapd/connection.c b/servers/slapd/connection.c
index c739614fe..44ec5b42e 100644
--- a/servers/slapd/connection.c
+++ b/servers/slapd/connection.c
@@ -734,6 +734,7 @@ static void connection_abandon( Connection *c )
 		LDAP_STAILQ_NEXT(o, o_next) = NULL;
 		slap_op_free( o, NULL );
 	}
+	c->c_n_ops_pending = 0;
 }
 
 static void
-- 
2.33.0

