From 95f959e2155420cdd274a9fc1f75011a075d11fc Mon Sep 17 00:00:00 2001
From: Howard Chu <hyc@openldap.org>
Date: Thu, 16 Jun 2022 16:10:59 +0100
Subject: [PATCH] ITS#9864 slapo-accesslog: plug onetime memleaks

---
 servers/slapd/overlays/accesslog.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/servers/slapd/overlays/accesslog.c b/servers/slapd/overlays/accesslog.c
index 49a36df59..5418a66e3 100644
--- a/servers/slapd/overlays/accesslog.c
+++ b/servers/slapd/overlays/accesslog.c
@@ -2448,6 +2450,8 @@ accesslog_db_destroy(
 		ch_free( li->li_sids );
 	if ( li->li_mincsn )
 		ber_bvarray_free( li->li_mincsn );
+	if ( li->li_db_suffix.bv_val )
+		ch_free( li->li_db_suffix.bv_val );
 	ldap_pvt_thread_mutex_destroy( &li->li_log_mutex );
 	ldap_pvt_thread_mutex_destroy( &li->li_op_rmutex );
 	free( li );
-- 
2.33.0

