From cc24cf620470e600d31fd68f63decae82b9745f3 Mon Sep 17 00:00:00 2001
From: Howard Chu <hyc@openldap.org>
Date: Wed, 2 May 2018 16:51:49 +0100
Subject: [PATCH 011/109] ITS#8843 check for NULL modlist

---
 servers/slapd/overlays/syncprov.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/servers/slapd/overlays/syncprov.c b/servers/slapd/overlays/syncprov.c
index 684421fb9..9051ec9c9 100644
--- a/servers/slapd/overlays/syncprov.c
+++ b/servers/slapd/overlays/syncprov.c
@@ -1924,6 +1924,7 @@ syncprov_op_response( Operation *op, SlapReply *rs )
 		/* Don't do any processing for consumer contextCSN updates */
 		if ( SLAPD_SYNC_IS_SYNCCONN( op->o_connid ) &&
 			op->o_tag == LDAP_REQ_MODIFY &&
+			op->orm_modlist &&
 			op->orm_modlist->sml_op == LDAP_MOD_REPLACE &&
 			op->orm_modlist->sml_desc == slap_schema.si_ad_contextCSN ) {
 			/* Catch contextCSN updates from syncrepl. We have to look at
-- 
2.19.1

